lib/nerves_hub_web/channels/device_channel/messages.ex

```diff
@@ -0,0 +1,135 @@
+defmodule NervesHubWeb.DeviceChannel.Messages do
+  @moduledoc false
+  alias NervesHub.Devices.Device
+
+  require Logger
+  @type alarm_id() :: String.t()
+  @type alarm_description() :: String.t()
+
+  @type health_check_report() :: %{
+          timestamp: DateTime.t(),
+          metadata: %{String.t() => String.t()},
+          alarms: %{alarm_id() => alarm_description()},
+          metrics: %{String.t() => number()},
+          checks: %{String.t() => %{pass: boolean(), note: String.t()}}
+        }
+
+  @type scripts_run() :: %{
+          ref: String.t(),
+          output: String.t(),
+          return: String.t()
+        }
+
+  @type fwup_progress() :: %{
+          percent: integer()
+        }
+
+  @type location() :: term()
+
+  @type connection_types() :: %{types: list(atom())}
+
+  @type status_update() :: map()
+
+  @type check_update_available() :: map()
+
+  # We parse out messages explicitly to let the compiler help with types and
+  # to keep track of what we have coming in and out of the system
+  # They are not structs to reduce the proliferation of modules for what is mostly
+  # an inbetween layer
+  # If the role of these definitions grows to much it may make sense to turn them into
+  # structs.
+  @spec parse(event :: String.t(), params :: map()) ::
+          {:fwup_progress, fwup_progress()}
+          | {:location_update, location()}
+          | {:connection_types, connection_types()}
+          | {:status_update, status_update()}
+          | {:check_update_available, check_update_available()}
+          | {:health_check_report, health_check_report()}
+          | {:scripts_run, scripts_run()}
+          | {:rebooting, map()}
+          | {:unknown, map()}
+  def parse(event, params)
+
+  def parse("fwup_progress", %{"value" => percent}) do
+    {:fwup_progress, %{percent: percent}}
+  end
+
+  def parse("location:update", location) do
+    {:location_update, location}
+  end
+
+  @valid_types Device.connection_types()
+  def parse("connection_types", %{"values" => types}) do
+    types =
+      types
+      |> Enum.map(fn type ->
+        try do
+          String.to_existing_atom(type)
+        rescue
+          _ -> nil
+        end
+      end)
+      |> Enum.filter(fn type ->
+        if type in @valid_types do
+          true
+        else
+          Logger.warning("Received invalid type for connection_types: #{inspect(type)}")
+          false
+        end
+      end)
```

Comment:
```suggestion
      for {type, type_str} <- Ecto.Enum.mappings(Device, :connection_types), type_str in types, do: type
```
We can use the Ecto.Enum mappings to get the valid types out of this. I'm unsure we need to warn on invalid being included? (this would mostly be hidden to the users anyway)

Reply:
Just figured we'd do something when we see weird things :)

Would Ecto.Enum be smoother somehow or are you thinking full-on embedded schema?